### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by teatreejasmine**
Test

### CODE REVIEW

File: StackOverflow/Controller/QuestionsCell.h [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+//
+//  Created by Mali on 2017/10/04.
+//  Copyright © 2017 DVT. All rights reserved.
+//
+
+#import <UIKit/UIKit.h>
+
+@interface QuestionsCell : UITableViewCell
+
+@property (weak, nonatomic) IBOutlet UIImageView *acceptedAnswerImage;
```
Comment (vincent-coetzee) [inline StackOverflow/Controller/QuestionsCell.h]: Naming is very important, and as a code reviewer I'll always spend a considerable amount of time on naming. So in this case, the name is incorrect. It's not an image, it's an imageView and therefore this should be named acceptedAnswerImageView.

File: StackOverflow/Controller/QuestionsCell.h [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+@property (weak, nonatomic) IBOutlet UILabel *questionTitle;
```
Comment (vincent-coetzee) [inline StackOverflow/Controller/QuestionsCell.h]: Similarly this is not a questionTitle, it's a questionTitleLabel.

File: StackOverflow/Controller/QuestionsCell.h [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+@property (weak, nonatomic) IBOutlet UILabel *tags;
```
Comment (vincent-coetzee) [inline StackOverflow/Controller/QuestionsCell.h]: Similarly this should be tagsLabel

File: StackOverflow/Controller/QuestionsCell.h [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+@property (weak, nonatomic) IBOutlet UILabel *answerCount;
```
Comment (vincent-coetzee) [inline StackOverflow/Controller/QuestionsCell.h]: answerCountLabel. Remember the Swift documentation about names being meaningful at the point of use. If I look at this I think it contains an int or a float value because it's a count, but actually it's not, it's a label. You MUST get your naming correct. If PRs have bad naming I will reject them.

File: StackOverflow/Controller/QuestionsCell.m [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+//  QuestionsCell.m
+//  StackOverflow
+//
+//  Created by Mali on 2017/10/04.
+//  Copyright © 2017 DVT. All rights reserved.
+//
+
+#import "QuestionsCell.h"
+
+@implementation QuestionsCell
```
Comment (vincent-coetzee) [inline StackOverflow/Controller/QuestionsCell.m]: The logic for setting the contents of a cell should be contained in the TableViewCell subclass. If you need me to explain what I mean here please let me know. This is a common error that many developers make. The code for configuring the content of TableViewCells is contained in the UITableViewDataSource which is incorrect for a number of reasons. Ideally you should define a single setter on the TableViewCell subclasses that takes the model that the TableViewCell subclasses represents and in that setter all the config of the TableViewCell should be done.

File: StackOverflow/Controller/QuestionsCell.m [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    [super awakeFromNib];
+    // Initialization code
+}
+
+- (void)setSelected:(BOOL)selected animated:(BOOL)animated {
+    [super setSelected:selected animated:animated];
+
+    // Configure the view for the selected state
+}
+
```
Comment (vincent-coetzee) [inline StackOverflow/Controller/QuestionsCell.m]: Why is prepareForReuse not implemented ?

File: StackOverflow/Controller/QuestionsTableViewController.m [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+#import "QuestionsTableViewController.h"
+#import "QuestionsCell.h"
+#import "DataFetchDelegate.h"
+#import "QuestionsDataObject.h"
+#import "QuestionsData.h"
+
+
+@interface QuestionsTableViewController () <DataFetchDelegate> {
+    NSArray *dataResult;
```
Comment (vincent-coetzee) [inline StackOverflow/Controller/QuestionsTableViewController.m]: This is neither data nor a result. Variables should be named after their role not their type. What role does this array fulfil ? Is it a list of something, then call it that. Calling something dataResult is meaningless. Also please be explicit when defining the storage modifiers of variables, in this case I assume you want this variable to have a strong reference ( which is the default ) therefore annotate this with __strong so that other developers know that is what you mean. ALWAYS be explicit when annotating variables and properties.

File: StackOverflow/Controller/QuestionsTableViewController.m [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    QuestionsDataObject *dataObject;
```
Comment (vincent-coetzee) [inline StackOverflow/Controller/QuestionsTableViewController.m]: Variable names should never end in Object, we know they are objects, in Objective C they can not be anything else. Once again remember your naming, variable must be named after their role, NOT their type. Data has a very specific meaning in iOS, it means an instance of NSData or in Swift a struct of type Data. If I see a name with data in it, I expect it to be one of these types not some arbitrary object. What is this ? Is it a Question, if it is a Question then call it that. If it is a list of Questions then call it listOfQuestions. This name does not actually telling me anything.

File: StackOverflow/Controller/QuestionsTableViewController.m [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+
+- (void)viewDidLoad {
+    [super viewDidLoad];
+    
+    //Display a code loading indicator
+    activityIndicator = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleGray];
+    activityIndicator.alpha = 1.0;
+    [self.view addSubview:activityIndicator];
+    activityIndicator.center = CGPointMake([[UIScreen mainScreen]bounds].size.width/2, [[UIScreen mainScreen]bounds].size.height/2);
```
Comment (vincent-coetzee) [inline StackOverflow/Controller/QuestionsTableViewController.m]: Views should be positioned using Auto Layout not by directly setting their location.

File: StackOverflow/Controller/QuestionsTableViewController.m [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    [activityIndicator startAnimating];
+
+    dataObject = [[QuestionsDataObject alloc] init];
+    dataObject.questionRequest = [[QuestionRequest alloc] init];
+    dataObject.questionRequest.delegate = dataObject;
+    dataObject.delegate = self;
+    
+    [dataObject getData];
```
Comment (vincent-coetzee) [inline StackOverflow/Controller/QuestionsTableViewController.m]: Try to make your names more meaningful. Get what data from where ? Invocations on Objective C should read like sentences. This does not, do you see here how meaningless dataObject as a name is ? It tells me nothing. If the QuestionsDataObject is actually an interface to a backend question server then call it that. Then this should read something like [remoteQuestionServer fetchListOfQuestions] or something like that. It tells me more about what is going on.

File: StackOverflow/Controller/QuestionsTableViewController.m [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
+    return dataResult.count;
+}
+
+- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
+    QuestionsCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
+    
+    QuestionsData *questionData = dataResult[indexPath.row];
+    
+    //Select image based on whether the answer is accepted or not
```
Comment (vincent-coetzee) [inline StackOverflow/Controller/QuestionsTableViewController.m]: The configuration of the presentation performed by the QuestionCell should not be done inside the DataSource for several reasons. Firstly, the state and layout of the QuestionCell is private to that cell and therefore you violate encapsulation ( a primary tenet of OO ) by configuring the cell inside the DataSource implementation. Secondly if you ever reuse the QuestionCell in another controller, you will have to recode all the configuration logic in that DataSource as well. Define a setter on the QuestionCell that accepts the question, save the model ( i.e. the question ) in the cell, and then in that setter call another method on the QuestionCell that actually configures the presentation from the stored question. If you need me to explain this please contact me.

File: StackOverflow/Model/DataDelegate.h [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+//
+//  Created by Malea on 2017/10/04.
+//  Copyright © 2017 DVT. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+
+@protocol DataDelegate <NSObject>
+
+- (void)receivedDataJSON:(NSData *)objectNotation;
```
Comment (vincent-coetzee) [inline StackOverflow/Model/DataDelegate.h]: These should be defined as required since the delegate can't function as a delegate without them.

File: StackOverflow/Model/DataDelegate.h [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+//  DataDelegate.h
+//  StackOverflow
+//
+//  Created by Malea on 2017/10/04.
+//  Copyright © 2017 DVT. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+
+@protocol DataDelegate <NSObject>
```
Comment (vincent-coetzee) [inline StackOverflow/Model/DataDelegate.h]: Once again naming, DataDelegate is so generic it does not tell me anything. What sort of data is this ? Is it questions, is it answers what ? Name things to clearly convey their role.

File: StackOverflow/Model/DataFetchDelegate.h [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+//  DataFetchDelegate.h
+//  StackOverflow
+//
+//  Created by Malea on 2017/10/04.
+//  Copyright © 2017 DVT. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+
+@protocol DataFetchDelegate <NSObject>
```
Comment (vincent-coetzee) [inline StackOverflow/Model/DataFetchDelegate.h]: I'm not sure why we actually need two protocols, surely one would have been sufficient ?

File: StackOverflow/Model/DataFetchDelegate.h [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+
+- (void)didReceiveData:(NSArray *)groups;
```
Comment (vincent-coetzee) [inline StackOverflow/Model/DataFetchDelegate.h]: Why is the method called didReceiveData: when in fact it's groups ? Surely it should be called didReceiveQuestionGroups ?

File: StackOverflow/Model/DataResult.h [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+//  DataResult.h
+//  StackOverflow
+//
+//  Created by Malea on 2017/10/04.
+//  Copyright © 2017 DVT. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+
+@interface DataResult : NSObject
```
Comment (vincent-coetzee) [inline StackOverflow/Model/DataResult.h]: I've already commented on this class' name !

File: StackOverflow/Model/DataResult.h [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+
++ (NSArray *)resultFromJSON:(NSData *)objectNotation error:(NSError **)error;
```
Comment (vincent-coetzee) [inline StackOverflow/Model/DataResult.h]: resultFromJSON is not a good name, a result is too generic. What are we actually getting from the JSON, questions, answers, groups ?

File: StackOverflow/Model/DataResult.m [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+    NSArray *results = [parsedObject valueForKey:@"items"];
+    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:results options:NSJSONWritingPrettyPrinted error:error];
+    NSString *jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
+    
+    NSError *err = nil;
+    NSArray *array = [NSJSONSerialization JSONObjectWithData:[jsonString dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingMutableContainers error:&err];
+
+    for (NSDictionary *questionDictionary in array) {
+        QuestionsData *questionData = [[QuestionsData alloc] init];
+
```
Comment (vincent-coetzee) [inline StackOverflow/Model/DataResult.m]: What are the blank lines for ? They add no value and merely make the code difficult to read. When using whitespace be judicious, if you just add blank lines everywhere they stop adding value.

File: StackOverflow/Model/DataResult.m [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+    NSArray *array = [NSJSONSerialization JSONObjectWithData:[jsonString dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingMutableContainers error:&err];
+
+    for (NSDictionary *questionDictionary in array) {
+        QuestionsData *questionData = [[QuestionsData alloc] init];
+        
+        for (NSString *key in questionDictionary) {
+
+            if ([questionData respondsToSelector:NSSelectorFromString(key)]) {
+                [questionData setValue:[questionDictionary valueForKey:key] forKey:key];
+
```
Comment (vincent-coetzee) [inline StackOverflow/Model/DataResult.m]: As above

File: StackOverflow/Model/DataResult.m [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+    NSString *jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
+    
+    NSError *err = nil;
+    NSArray *array = [NSJSONSerialization JSONObjectWithData:[jsonString dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingMutableContainers error:&err];
+
+    for (NSDictionary *questionDictionary in array) {
+        QuestionsData *questionData = [[QuestionsData alloc] init];
+        
+        for (NSString *key in questionDictionary) {
+
```
Comment (vincent-coetzee) [inline StackOverflow/Model/DataResult.m]: As above

File: StackOverflow/Model/QuestionRequest.m [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+//  Copyright © 2017 DVT. All rights reserved.
+//
+
+#import "QuestionRequest.h"
+#import "DataDelegate.h"
+
+@implementation QuestionRequest
+
+- (void)fetchData {
+
```
Comment (vincent-coetzee) [inline StackOverflow/Model/QuestionRequest.m]: Meaningless blank line

File: StackOverflow/Model/QuestionRequest.m [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+#import "QuestionRequest.h"
+#import "DataDelegate.h"
+
+@implementation QuestionRequest
+
+- (void)fetchData {
+    
+    NSString *urlAsString = [NSString stringWithFormat:@"https://api.stackexchange.com/2.2/questions?pagesize=50&order=desc&sort=creation&tagged=ios&site=stackoverflow"];
+
```
Comment (vincent-coetzee) [inline StackOverflow/Model/QuestionRequest.m]: Blank line ?

File: StackOverflow/Model/QuestionRequest.m [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+#import "DataDelegate.h"
+
+@implementation QuestionRequest
+
+- (void)fetchData {
+    
+    NSString *urlAsString = [NSString stringWithFormat:@"https://api.stackexchange.com/2.2/questions?pagesize=50&order=desc&sort=creation&tagged=ios&site=stackoverflow"];
+    
+    NSURL *url = [[NSURL alloc] initWithString:urlAsString];
+
```
Comment (vincent-coetzee) [inline StackOverflow/Model/QuestionRequest.m]: And again

File: StackOverflow/Model/QuestionRequest.m [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+- (void)fetchData {
+    
+    NSString *urlAsString = [NSString stringWithFormat:@"https://api.stackexchange.com/2.2/questions?pagesize=50&order=desc&sort=creation&tagged=ios&site=stackoverflow"];
+    
+    NSURL *url = [[NSURL alloc] initWithString:urlAsString];
+  
+    NSURLSession *session = [NSURLSession sharedSession];
+    [[session dataTaskWithURL:url completionHandler:^(NSData *data, NSURLResponse *response, NSError *error){
+        if (error) {
+            [self.delegate fetchingDataFailedWithError:error];
```
Comment (vincent-coetzee) [inline StackOverflow/Model/QuestionRequest.m]: I mentioned earlier on that you should define the methods on your delegate as required. The way this code stands it could crash because the delegate could be non nil and it could not understand these methods. There are two ways to correct this, either test to see whether the delegate actually implements the method or make the methods in the protocol required. Good practice also suggests that you should actually test to see whether the delegate is nil or not. This code will work on OC but it could crash in Swift.

File: StackOverflow/Model/QuestionsData.h [hunk 25]
```diff
@@ -0,0 +1,18 @@
...
+//  Copyright © 2017 DVT. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+
+@interface QuestionsData : NSObject
+
+@property (strong, nonatomic) NSString *title;
+@property (strong, nonatomic) NSArray *tags;
+@property (strong, nonatomic) NSString *answer_count;
```
Comment (vincent-coetzee) [inline StackOverflow/Model/QuestionsData.h]: We do NOT use _ ( or snake case ) in OC or Swift. We use camel case. i.e. answerCount.

File: StackOverflow/Model/QuestionsData.h [hunk 26]
```diff
@@ -0,0 +1,18 @@
...
+@property (strong, nonatomic) NSString *accepted_answer_id;
```
Comment (vincent-coetzee) [inline StackOverflow/Model/QuestionsData.h]: As above

File: StackOverflow/Model/QuestionsData.m [hunk 27]
```diff
@@ -0,0 +1,13 @@
...
+//  QuestionsData.m
+//  StackOverflow
+//
+//  Created by Malea on 2017/10/04.
+//  Copyright © 2017 DVT. All rights reserved.
+//
+
+#import "QuestionsData.h"
+
+@implementation QuestionsData
```
Comment (vincent-coetzee) [inline StackOverflow/Model/QuestionsData.m]: If this class has no behaviour do we actually need it, or is it a DTO, in which case it should be named as such.

File: StackOverflow/Model/QuestionsDataObject.h [hunk 28]
```diff
@@ -0,0 +1,28 @@
...
+#import "DataDelegate.h"
+#import "QuestionRequest.h"
+
+
+@class QuestionsRequest; 
+
+@interface QuestionsDataObject : NSObject <DataDelegate>
+
+@property (strong, nonatomic) QuestionRequest *questionRequest;
+@property (weak, nonatomic) id <DataFetchDelegate> delegate;
```
Comment (vincent-coetzee) [inline StackOverflow/Model/QuestionsDataObject.h]: It might be better to type this as "NSObject < DataFetchDelegate > ". As a rule id should be avoided as much as possible because it bypasses all the possible type checks the compiler can do.

File: StackOverflow/Model/QuestionsDataObject.h [hunk 29]
```diff
@@ -0,0 +1,28 @@
...
+
+- (void)getData;
```
Comment (vincent-coetzee) [inline StackOverflow/Model/QuestionsDataObject.h]: Needs a better name

File: StackOverflow/Model/QuestionsDataObject.m [hunk 30]
```diff
@@ -0,0 +1,34 @@
...
+//  Created by Malea on 2017/10/04.
+//  Copyright © 2017 DVT. All rights reserved.
+//
+
+#import "QuestionsDataObject.h"
+#import "DataResult.h"
+
+@implementation QuestionsDataObject
+
+- (void)getData {
```
Comment (vincent-coetzee) [inline StackOverflow/Model/QuestionsDataObject.m]: All this class seems to do is forward the messages to the question request. If that's the case why is it needed ?

File: StackOverflowTests/StackOverflowTests.m [hunk 31]
```diff
@@ -0,0 +1,39 @@
...
+
+#import <XCTest/XCTest.h>
+
+@interface StackOverflowTests : XCTestCase
+
+@end
+
+@implementation StackOverflowTests
+
+- (void)setUp {
```
Comment (vincent-coetzee) [inline StackOverflowTests/StackOverflowTests.m:19]: Unit tests are always a good idea.

**Review by vincent-coetzee — CHANGES_REQUESTED** [review]
Hi Malea. Please contact me if you would like me to explain things further.

File: StackOverflow/Controller/QuestionsCell.h (hunk 1 above)
Comment (teatreejasmine) [reply StackOverflow/Controller/QuestionsCell.h]: Thank you, I will correct this

File: StackOverflow/Controller/QuestionsCell.h (hunk 4 above)
Comment (teatreejasmine) [reply StackOverflow/Controller/QuestionsCell.h]: I will correct this thank you

File: StackOverflow/Controller/QuestionsCell.m (hunk 5 above)
Comment (teatreejasmine) [reply StackOverflow/Controller/QuestionsCell.m]: Logic added in the TableViewCell
